com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/Nodes/HDSampleBufferNode.cs

```diff
@@ -0,0 +1,199 @@
+                    channelCount = 3;
+                    break;
+                case BufferType.Roughness:
+                    AddSlot(new Vector1MaterialSlot(k_OutputSlotId, k_OutputSlotName, k_OutputSlotName, SlotType.Output, 0, ShaderStageCapability.Fragment));
+                    channelCount = 1;
+                    break;
+                case BufferType.MotionVectors:
+                    AddSlot(new Vector2MaterialSlot(k_OutputSlotId, k_OutputSlotName, k_OutputSlotName, SlotType.Output, Vector2.zero, ShaderStageCapability.Fragment));
+                    channelCount = 2;
+                    break;
+                case BufferType.PostProcessInput:
+                    AddSlot(new ColorRGBAMaterialSlot(k_OutputSlotId, k_OutputSlotName, k_OutputSlotName, SlotType.Output, Color.black, ShaderStageCapability.Fragment));
+                    channelCount = 4;
+                    break;
+            }
+
+            RemoveSlotsNameNotMatching(new[]
+            {
+                k_ScreenPositionSlotId,
+                k_SamplerInputSlotId,
+                k_OutputSlotId,
+            });
+        }
+
+        string GetFunctionName() => $"Unity_HDRP_SampleBuffer_{bufferType}_$precision";
+
+        public void GenerateNodeFunction(FunctionRegistry registry, GenerationMode generationMode)
+        {
+            // Preview SG doesn't have access to HDRP depth buffer
+            if (!generationMode.IsPreview())
+            {
+                registry.ProvideFunction(GetFunctionName(), s =>
+                {
+                    s.AppendLine("#include \"Packages/com.unity.render-pipelines.high-definition/Runtime/Material/NormalBuffer.hlsl\"");
+                    s.AppendLine("#include \"Packages/com.unity.render-pipelines.high-definition/Runtime/Material/Builtin/BuiltinData.hlsl\"");
+
+                    // Declare post process input here because the property collector don't support TEXTURE_X type
+                    s.AppendLine($"TEXTURE2D_X({nameof(HDShaderIDs._CustomPostProcessInput)});");
+                    s.AppendLine($"SAMPLER(sampler{nameof(HDShaderIDs._CustomPostProcessInput)});");
+
+                    s.AppendLine("$precision{1} {0}($precision2 uv, SamplerState samplerState)", GetFunctionName(), channelCount);
+                    using (s.BlockScope())
+                    {
+                        switch (bufferType)
+                        {
+                            case BufferType.NormalWorldSpace:
+                                s.AppendLine("uint2 pixelCoords = uint2(uv * _ScreenSize.xy);");
+                                s.AppendLine("NormalData normalData;");
+                                s.AppendLine("DecodeFromNormalBuffer(pixelCoords, normalData);");
+                                s.AppendLine("float depth = LoadCameraDepth(pixelCoords);");
+                                s.AppendLine("return depth > 0 ? normalData.normalWS : 0;");
```

Comment:
Remark, this test assume invert Z (which is the case for hdrp) but to have in mind

Reply:
we should rather abstract with an "IsSky()" instead

Reply:
Also generally speaking, I think it would be important in shader graph to have a way to know if a pixel from a target( normal / roughness) is "valid". As here users have no way to know it other than redoing this exact same test with depth

Reply:
hmm... maybe currently this node (HDSampleBuffer), should have an extra output to say that input is valid or not

Reply:
I've added an IsSky function and an extra output for the node that return the IsSky() value